calrissian/report.py

```diff
@@ -0,0 +1,340 @@
+
+
+class TimedReport(object):
+    """
+    Report on operations with a specific start time and finish time.
+    """
+
+    def __init__(self, start_time=None, finish_time=None):
+        self.start_time = start_time
+        self.finish_time = finish_time
+
+    def start(self, start_time=None):
+        self.start_time = start_time if start_time else datetime.now()
+
+    def finish(self, finish_time=None):
+        self.finish_time = finish_time if finish_time else datetime.now()
+
+    def elapsed_seconds(self):
+        delta = self.finish_time - self.start_time
+        total_seconds = delta.total_seconds()
+        if total_seconds < 0:
+            raise ValueError('Negative time is not allowed: {}'.format(total_seconds))
+        else:
+            return total_seconds
+
+    def elapsed_hours(self):
+        return self.elapsed_seconds() / SECONDS_PER_HOUR
+
+    def to_dict(self):
+        result = dict(vars(self)) # Make sure we create a copy, otherwise writing to the dict overwrites the object
+        result['elapsed_hours'] = self.elapsed_hours()
+        result['elapsed_seconds'] = self.elapsed_seconds()
+        return result
+
+
+class ResourceParser(object):
+    """
+    Base class for converting Kubernetes resources (memory/CPU) from strings to reportable numbers
+    """
+    kind = None
+    url = None
+    suffixes = None
+
+    @classmethod
+    def parse(cls, value):
+        try:
+            for suffix, factor in cls.suffixes.items():
+                if value.endswith(suffix):
+                    return float(value.replace(suffix, '')) * factor
+            # No suffix, assume raw number
+            return float(value)
+        except Exception:
+            raise ValueError('Unable to parse \'{}\' as {}. See {}'.format(value, cls.kind, cls.url))
+
+
+class MemoryParser(ResourceParser):
+    """
+    Converts Kubernetes memory resource strings (e.g. 1Mi, 1G)to byte quantities
+    """
+    kind = 'memory'
+    url = 'https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#meaning-of-memory'
+    suffixes = {
+        'E': 1e18,
+        'P': 1e15,
+        'T': 1e12,
+        'G': 1e9,
+        'M': 1e6,
+        'K': 1e3,
+        'Ei': 2**60,
+        'Pi': 2**50,
+        'Ti': 2**40,
+        'Gi': 2**30,
+        'Mi': 2**20,
+        'Ki': 2**10,
+    }
+
+    @classmethod
+    def parse_to_megabytes(cls, value):
+        return (cls.parse(value) / cls.suffixes['M'])
+
+
+class CPUParser(ResourceParser):
+    """
+    Converts Kubernetes CPU resource strings (e.g. 2, 200m) to floating point CPU quantities.
+    """
+
+    kind = 'cpu'
+    url = 'https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#meaning-of-cpu'
+    suffixes = {
+        'm': 0.001
+    }
+
+
+class TimedResourceReport(TimedReport):
+    """
+    Adds CPU and memory usage to TimedReport, in order to calculate resource
+    usage over the duration of the timed report
```

Comment:
I wasn't sure if the CPU and memory values here were actual or requested.
Perhaps a comment to clarify that.